common/flatpak-instance.c

```diff
@@ -429,6 +435,354 @@ flatpak_instance_new_for_id (const char *id)
   return flatpak_instance_new (dir);
 }
 
+/*
+ * Try to reuse an existing per-app temporary /dev/shm or /tmp
+ * directory referenced by @link_path. Return %TRUE if we can.
+ *
+ * We have to be careful here, because @link_path might be left over
+ * from a previous boot, and it probably points into a directory like
+ * /dev/shm or /tmp, where an attacker might recreate our directories,
+ * for example as symbolic links to somewhere they control.
+ */
+static gboolean
+flatpak_instance_reuse_per_app_tmp_directory (const char *app_id,
```

Comment:
This is security-sensitive, please review carefully.

Reply:
In today's version of this branch I've added somewhat extensive test coverage for this function. If you create a directory owned by root before testing (see comments) then the only branch not covered is `fstat()` failing - I honestly don't know how to make that happen.

Reply:
This function is now part of #4214.